module/zfs/zthr.c

```diff
@@ -319,25 +411,17 @@ zthr_resume(zthr_t *t)
 boolean_t
 zthr_iscancelled(zthr_t *t)
 {
-	boolean_t cancelled;
-
 	ASSERT3P(t->zthr_thread, ==, curthread);
 
-	mutex_enter(&t->zthr_lock);
-	cancelled = t->zthr_cancel;
-	mutex_exit(&t->zthr_lock);
-
+	/*
+	 * Keep in mind that we are not holding the request_lock here
+	 * as this is called from the zthr_func callback and it wants
+	 * to check if we have an active zthr_cancel(). If that's the
+	 * case then that zthr_cancel() will be holding the request
+	 * lock.
```

Comment:
This comment is confusing. As a caller, I'm not sure if I should be holding request_lock or not.

Reply:
That makes more sense. I think some more clarification somewhere about the use of the request lock vs the state lock would be good. Perhaps up in the paragraph titled `== Implementation of ZTHR requests`. 

In general I'm just confused why we have 2 locks when the request lock is only ever taken before taking the state lock. Up until when you pointed out that the structure is defined within this file, I thought the request lock was meant to be held by callers in some situations. If its not, is the request lock really necessary?

Reply:
I'm copy pasting the paragraph here:
```
 * ZTHR wakeup, cancel, and resume are requests on a zthr to change
 * its internal state. Requests on a zthr are serialized using the
 * zthr_request_lock, while changes in its internal state are
 * protected by the zthr_state_lock. In general a request will first
 * acquire the zthr_request_lock to ensure that other requests can't
 * be served at the same time, and then will acquire the zthr_state_lock
 * to apply its changes. In cases like zthr_cancel() where we need
 * to coordinate the thread issuing the request and the zthr, zthr_cv
 * is used as the mechanism of communication.
```

The request lock is necessary because it is held by requests like
zthr_cancel() and zthr_resume() in order to get serialized. If we
only had the state lock, then the same race conditions that we had
before will come back. That is because zthr_cancel() has to drop the
state lock while doing a cv_wait(). Holding the request lock while
cv_waiting in zthr_cancel() we ensure that no other zthr request
takes place at the same time.

Is this making sense?

Reply:
Ok got it. What if we change everything in that paragraph after (and including) `In general a request ....` to:

```
A request will first acquire the zthr_request_lock and then immediately acquire the zthr_state_lock. We do this so that incoming requests are serialized using the request lock, while still allowing us to use the state lock for thread communication via zthr_cv.
```

Reply:
That's definitely more precise and clear. I will replace the existing part with your suggestion.